### ISSUE TITLE
Broken providers no_cache support

### ISSUE THREAD

**Issue by simo5** [openssl/openssl#26038]
Ok so this is a difficult to the bug one and apologies in advance that I couldn't find the actual triggering issue even after half a day of GDB sessions.

Here's what happens.

Configuration
==
the openssl.cnf file activates the default provider and the pkcs11-provider, no special options or evp parameters are provided.

Startup conditions
==
The pkcs11-provider does not have any special token configured, which means it will try to load a default token via p11-kit if explicitly invoked but will not really initialize anything.
In this case whenever ossl_provider_query_operation() is called for the provider it returns
a NULL pointer and *no_cache set to 1, to indicate this response should not be cache.

The problem
==
An application needs to use MD4 digesting (yeah I know), and does the following operations:

```
OSSL_PROVIDER_try_load(NULL, "legacy");
type = EVP_Md4();
ctx = EV_MD_CTX_new();
EVP_DigestInit_ex(tx, type, NULL);
```

Inside the init function EVP_MD_fetch (searching for "MD4" in this case) is attempted, trying to follow the method fetching calls I see that all three providers are found (default, pkcs11, legacy, and all three have their operation 1 (digesting) queried, but for reasons I cannot fully understand the code decides to not map any algorithm and a finally request to the algorithm store cahce fails to find anything and bails.

The working case
==
With the exact same code above but toggling a pkcs11-provider config option that lets it initialize early and therefore return pointers and *no_cache = 0 the EVP_MD_fetch resolves immediately from the alg store cache without even needing to construct a method (arguably the digests have been resolved by a previous call). The same behavior if the provider is disabled in config.

Conclusions so far
==
Somehow returning *no_cache = 1 (or perhaps NULL maps ?) confuses OpenSSL to the point that any attempt to resolve digest operations from any provider seem to fail, even if other providers do positively respond that they support the requested algorithm.

I tried to understand how all this caching code playes together but I could not find how/where in the code this could manifest. Hopefully someone in the core team understand the caching better.

In my opinion a misbheaving* provider should not cause other providers to fail to work so I think this is an OpenSSL bug regardless of whether the current pkcs11 provider behavior is correct or not.

Any hint on how to further debug this problem are also welcome.

* I think it should be allowed to defer mapping, int he pkcs11 provider case this is becasuse until a token is inserted we do not know what mechanisms the token supports and therefore we attempt to delay responding anything to OpenSSL until (a) we have a token, and (b) we know which mechanisms it supports so we can definitely return what we can do.

**Comment by t8m** [comment]
> * I think it should be allowed to defer mapping, int he pkcs11 provider case this is becasuse until a token is inserted we do not know what mechanisms the token supports and therefore we attempt to delay responding anything to OpenSSL until (a) we have a token, and (b) we know which mechanisms it supports so we can definitely return what we can do.

Just out of curiosity - how do you handle the case where the inserted token is exchanged for a different token that supports different mechanisms?

**Comment by simo5** [comment]
> Just out of curiosity - how do you handle the case where the inserted token is exchanged for a different token that supports different mechanisms?

We do not support (yet) token insertion/removal, which is also why we defer initialization at this time.

Once insertion/removal will be considered .. well I am afraid I will have to request an interface to invalidate caches.

The other option is to change pkcs11-provider to always claim everything is supported, but this comes with other problems: how to handle operations before tokens are available or operation the token does not support (like some specific digest).

It definitely would help to have a way to interact with the default libctx to be able to tell OpenSSL things have changed.

**Comment by simo5** [comment]
Additional info, configuring the legacy provider in the config file makes the code work, so we can narrow the conditions down to "providers only loaded after initial crypto init".

**Comment by simo5** [comment]
Ok I was able to narrow down one more thing.

Turned out the system I was testing this on had no token actually available this caused the actual provider initialization to fail.

Now with the pkcs11 provider option to initialize early the provider fails immediately at ossl_provider_init() and "openssl list -providers" doesn't even show the provider.

However without that option "openssl list -providers" shows the pkcs11 in the list (because pkcs11-provider delays internal initialization.

The difference is that if a valid token is inserted a delayed initialization seem not to cause the bug to happen. It is unclear to me why that is, I can only assume some operation in the application I am testing (wpa_supplicant) before the legacy provider use is attempted causes the pkcs11 provider to be fully initialized before operations are cached ...

**Comment by t8m** [comment]
> Once insertion/removal will be considered .. well I am afraid I will have to request an interface to invalidate caches.

Yeah, that might be much better interface than the `no_cache` thing.

What happens if you do not set `*no_cache = 1`? Is pkcs11 provider never queried again?

**Comment by simo5** [comment]
@t8m because we return no algorithms if we return *no_cache = 0 then the provider methods will never again be available, indeed.

**Comment by simo5** [comment]
apparently openssl does also negative caching?

**Comment by beldmit** [comment]
@simo5 I think you mentioned some other details to reproduce/avoid the buggy behavior

**Comment by simo5** [comment]
FWIW, I think this should be fixed in current versions, a provider that does not behave as intended should not affect other providers in any case (IMO), but for the long term, I think OpenSSL should provide new providers with a callback they can invoke to ask OpenSSL to forget everything it knows about the provider and re-probe all names, this would allows pkc11-provider to properly handle late initialization and token (re)insertion.

**Comment by beldmit** [comment]
@simo5 I can't reproduce the issue with the minimal reproducer you propose. pkcs11-provider 0.6-2, I've uncommented the currently commented activation

**Comment by beldmit** [comment]
Using the current (quite inconvenient) reproducer, we get  ERR_R_UNSUPPORTED in `inner_evp_generic_fetch`

**Comment by t8m** [comment]
@beldmit if you have the reproducer readily available, could you please try to modify the pkcs11 provider to always set *no_cache = 1 even if it has a PKCS11 module loaded? And then try with a module loaded i.e., so it does not return empty query results. Is the failure of the fetch of MD4 from legacy stil reproducible in such conditions? I.e. is the problem with no_cache being returned or is it rather a problem with empty query results?

**Comment by beldmit** [comment]
So. I have a simple cmdline reproducer

```
HOME			= .

# Use this in order to automatically load providers.
openssl_conf = openssl_init

# Comment out the next line to ignore configuration errors
config_diagnostics = 1

[openssl_init]
providers = provider_sect

[provider_sect]
default = default_sect
pkcs11 = pkcs11_sect
oqsprovider = oqs_sect

[default_sect]
activate = 1

[oqs_sect]
activate = 1

[pkcs11_sect]
activate = 1
```

```
$ OPENSSL_CONF=multiprov.cnf openssl genpkey -algorithm mlkem1024 -outform DER -out prkey.der -outpubkey pubkey.der
Error writing key(s)
80C2B598A07F0000:error:1D800065:ENCODER routines:OSSL_ENCODER_to_bio:reason(101):crypto/encode_decode/encoder_lib.c:55:No encoders were found. For standard encoders you need at least one of the default or base providers available. Did you forget to load them?
80C2B598A07F0000:error:1D800065:ENCODER routines:OSSL_ENCODER_to_bio:reason(101):crypto/encode_decode/encoder_lib.c:55:No encoders were found. For standard encoders you need at least one of the default or base providers available. Did you forget to load them?
80C2B598A07F0000:error:068000C4:asn1 encoding routines:i2d_provided:unsupported type:crypto/asn1/i2d_evp.c:69:
```

**Comment by beldmit** [comment]
If you change the order and place oqsprovider _before_ pkcs11 provider, everything works like a charm

**Comment by beldmit** [comment]
Hmmm. Even more strange.
The command: 
`$ OPENSSL_CONF=multiprov.cnf openssl genpkey -algorithm mlkem1024 -outform DER -out prkey.der -outpubkey pubkey.der`

This config works:
```
HOME                    = .

# Use this in order to automatically load providers.
openssl_conf = openssl_init

# Comment out the next line to ignore configuration errors
config_diagnostics = 1

[openssl_init]
providers = provider_sect

[provider_sect]
default = default_sect
pkcs11 = pkcs11_sect
oqsprovider = oqs_sect
legacy = legacy_sect

[default_sect]
activate = 1

[oqs_sect]
activate = 1

[pkcs11_sect]
activate = 1

[legacy_sect]
activate = 1
```
This doesn't
```
HOME			= .

# Use this in order to automatically load providers.
openssl_conf = openssl_init

# Comment out the next line to ignore configuration errors
config_diagnostics = 1

[openssl_init]
providers = provider_sect

[provider_sect]
default = default_sect
pkcs11 = pkcs11_sect
legacy = legacy_sect
oqsprovider = oqs_sect

[default_sect]
activate = 1

[oqs_sect]
activate = 1

[pkcs11_sect]
activate = 1

[legacy_sect]
activate = 1
```
```

**Comment by beldmit** [comment]
Reproduced with the current master branch

**Comment by beldmit** [comment]
FTR: with encoders/decoders PKCS11 provider always returns *no_cache = 1

**Comment by t8m** [comment]
@beldmit What I am trying to see is whether the empty query results with *no_cache =1 is a necessary condition or not.

**Comment by beldmit** [comment]
I hardcoded `*no_cache = 1` for `OSSL_OP_ENCODER`

```
[pkcs11_sect]
activate = 1
#pkcs11-module-load-behavior = early
pkcs11-module-path = /usr/lib64/opensc-pkcs11.so
``` 
causes failure.

```
[pkcs11_sect]
activate = 1
pkcs11-module-load-behavior = early
pkcs11-module-path = /usr/lib64/opensc-pkcs11.so
```
does not

**Comment by beldmit** [comment]
Comparing the TRACE logs for success and failure, I see many lines like
`TRACE[XX:XX...]:PROVIDER: (provider pkcs11) query_operation failed`
that don't happen when we succeed

**Comment by beldmit** [comment]
One more: we always return NULL (not an empty array) and *no_cache = 1

**Comment by beldmit** [comment]
If we initialize the token early, we get a relevant value for algos and still get a failure.

**Comment by beldmit** [comment]
> @beldmit What I am trying to see is whether the empty query results with *no_cache =1 is a necessary condition or not.

@t8m Looks like it is.
Please note that the current reproducer is not about MD4 but about encoder/decoder.

**Comment by beldmit** [comment]
When we enforce `*no_cache=1` we get a failure both with and without early load.
`*no_cache = ctx->status == P11PROV_UNINITIALIZED ? 1 : 0;` succeeds with early load and fails without

**Comment by simo5** [comment]
This means returning NULL seem not a necessary condition, because with early load all the operations are allocated and therefore we return an actual pointer.

**Comment by beldmit** [comment]
And probably even simpler reproducer

`OPENSSL_MODULES=/usr/lib64/ossl-modules/ OPENSSL_CONF=multiprov.cnf ./util/wrap.pl apps/openssl list --encoders`

**Comment by beldmit** [comment]
`./config shared disable-cached-fetch` resolves the issue. 

That's enough for Friday :)

**Comment by beldmit** [comment]
I strongly suspect that calling `ossl_method_construct` always with 0 is a wrong thing.

**Comment by t8m** [comment]
> `./config shared disable-cached-fetch` resolves the issue.
> 
> That's enough for Friday :)

That's quite understandable as it basically makes all query calls non-cached.

**Comment by beldmit** [comment]
2 providers are enough, no need to deal with oqs provider
cat multiprov.cnf
```
openssl_conf = openssl_init

[openssl_init]
providers = provider_sect

[provider_sect]
pkcs11 = pkcs11_sect
default = default_sect

[default_sect]
activate = 1

[pkcs11_sect]
activate = 1
#mkdir /tmp/mytoken
#echo "12345678" > /tmp/mytoken.txt
#certutil -N -d /tmp/mytoken -f /tmp/mytoken.txt
pkcs11-module-path = /usr/lib64/libsoftokn3.so
pkcs11-module-init-args = configDir=/tmp/mytoken
#pkcs11-module-load-behavior = early
```
```
$ OPENSSL_MODULES=/usr/lib64/ossl-modules/ OPENSSL_CONF=multiprov.cnf ./util/wrap.pl apps/openssl list --encoders
Provided ENCODERs:
```

If we swap order of provider loading, everything starts working

**Comment by beldmit** [comment]
Resolved by #26197, backports to 3.0 are in #26231 and #26232

**Comment by beldmit** [comment]
@mouse07410 I think I've merged it to 3.2+
3.0/3.1 require a backport and have separate PRs
